cpp/include/Ice/Connection.h

```diff
@@ -144,11 +123,26 @@ namespace Ice
         virtual ~Connection() = 0;
 
         /**
-         * Manually close the connection using the specified closure mode.
-         * @param mode Determines how the connection will be closed.
-         * @see ConnectionClose
+         * Aborts this connection.
+         */
+        virtual void abort() noexcept = 0;
+
+        /**
+         * Starts a graceful closure of this connection once all outstanding invocations have completed.
+         * @param whenClosed A callback that is called when the connection is closed. Its parameter is always non-
+         * null and describes the reason for the closure. Several exceptions, including Ice::ConnectionClosedException,
+         * indicate a graceful closure. This callback is called immediately (and synchronously) if close is called on a
+         * closed connection. You can pass nullptr if you don't need to be notified when the connection is closed.
+         * @remarks If this operation takes longer than the configured close timeout, the connection is aborted with a
+         * CloseTimeoutException.
+         */
+        virtual void close(std::function<void(std::exception_ptr)> whenClosed) noexcept = 0;
+
+        /**
+         * Starts a graceful closure of this connection once all outstanding invocations have completed.
+         * @return A future that completes then the connection is closed.
          */
-        virtual void close(ConnectionClose mode) noexcept = 0;
+        std::future<void> close() noexcept;
```

Comment:
The callback API is not really practical to use. That's why I provided this future overload.